File: web/src/memory/linear_memory.js
/**
 * TRACE: Linear Recurrent Memory Core (TypeScript Twin)
 * ====================================================
 *
 * Verified exact equivalent of core/linear_memory.py.
 * Single mathematical contract tested against experiments/expected_results.json.
 *
 * Primary Source:
 * BDH-CQ (Engdahl, Kosowski, Chorowski, Stamirowska et al., 2026, arXiv:2608.09888), §3.2:
 * Instantiates the additive special case S_t = S_{t-1} + v_t k_t^T.
 *
 * Explicit Non-Claim:
 * This educational toy model does not reconstruct BDH-CQ's proprietary U_θ function.
 */
export class LinearMemory {
    d;
    state;
    carrier;
    writeHistory = [];
    constructor(d, carrier) {
        if (d < 1) {
            throw new Error(`Dimension d must be >= 1, got ${d}`);
        }
        this.d = d;
        this.state = Array.from({ length: d }, () => Array(d).fill(0));
        if (carrier && carrier.length === d) {
            const norm = Math.hypot(...carrier);
            this.carrier = carrier.map((x) => x / norm);
        }
        else {
            const c = Array(d).fill(1);
            const norm = Math.hypot(...c);
            this.carrier = c.map((x) => x / norm);
        }
    }
    reset() {
        this.state = Array.from({ length: this.d }, () => Array(this.d).fill(0));
        this.writeHistory = [];
    }
    encodeScalarValue(val) {
        return this.carrier.map((c) => c * val);
    }
    decodeScalarValue(vec) {
        return vec.reduce((sum, v, i) => sum + v * this.carrier[i], 0);
    }
    write(key, value, label) {
        if (key.length !== this.d) {
            throw new Error(`Key dimension ${key.length} does not match state dimension ${this.d}`);
        }
        let v;
        let scalarVal;
        if (typeof value === "number") {
            v = this.encodeScalarValue(value);
            scalarVal = value;
        }
        else {
            if (value.length !== this.d) {
                throw new Error(`Value dimension ${value.length} does not match state dimension ${this.d}`);
            }
            v = value;
            scalarVal = this.decodeScalarValue(v);
        }
        // Outer product update: S_t = S_{t-1} + v k^T
        for (let r = 0; r < this.d; r++) {
            for (let c = 0; c < this.d; c++) {
                this.state[r][c] += v[r] * key[c];
            }
        }
        this.writeHistory.push({
            label: label || `step_${this.writeHistory.length}`,
            scalarValue: scalarVal,
            key: [...key],
            stateSnapshot: this.state.map((row) => [...row]),
        });
        return this.state.map((row) => [...row]);
    }
    read(kQuery) {
        if (kQuery.length !== this.d) {
            throw new Error(`Query key dimension ${kQuery.length} does not match state dimension ${this.d}`);
        }
        let denom = 0;
        for (let i = 0; i < this.d; i++) {
            denom += kQuery[i] * kQuery[i];
        }
        if (denom === 0) {
            throw new Error("Query key norm is zero.");
        }
        // S * kQuery
        const numer = Array(this.d).fill(0);
        for (let r = 0; r < this.d; r++) {
            let sum = 0;
            for (let c = 0; c < this.d; c++) {
                sum += this.state[r][c] * kQuery[c];
            }
            numer[r] = sum;
        }
        return numer.map((val) => val / denom);
    }
    readScalar(kQuery) {
        const vec = this.read(kQuery);
        return this.decodeScalarValue(vec);
    }
    static interferenceScore(kTarget, kInterferer) {
        let dot = 0;
        let denom = 0;
        for (let i = 0; i < kTarget.length; i++) {
            dot += kTarget[i] * kInterferer[i];
            denom += kTarget[i] * kTarget[i];
        }
        if (denom === 0) {
            throw new Error("Target key norm is zero.");
        }
        return dot / denom;
    }
}
/**
 * Deterministic PRNG (Mulberry32)
 */
function mulberry32(seed) {
    let s = seed >>> 0;
    return function () {
        s = (s + 0x6d2b79f5) >>> 0;
        let t = Math.imul(s ^ (s >>> 15), 1 | s);
        t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
        return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
}
/**
 * Standard Normal N(0, 1) generator via Box-Muller transform
 */
function nextGaussian(rng) {
    let u1 = 0;
    let u2 = 0;
    while (u1 === 0)
        u1 = rng();
    while (u2 === 0)
        u2 = rng();
    return Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
}
/**
 * Constructs a codebook mapping symbols to unit key vectors in R^d:
 * - If symbols.length <= d: constructs an orthonormal codebook via Gram-Schmidt QR decomposition.
 *   Zero cross-key interference by construction (k_i · k_j = 0 for all i != j).
 * - If symbols.length > d: constructs random unit vectors on S^(d-1).
 *   Vectors are forced to overlap because exact orthogonality is impossible in R^d.
 */
export function buildCodebook(symbols, d, seed = 42) {
    const numSymbols = symbols.length;
    const rng = mulberry32(seed);
    const codebook = {};
    if (numSymbols <= d) {
        // Generate d linearly independent gaussian column vectors in R^d
        const rawCols = [];
        for (let c = 0; c < d; c++) {
            const col = [];
            for (let r = 0; r < d; r++) {
                col.push(nextGaussian(rng));
            }
            rawCols.push(col);
        }
        // Gram-Schmidt orthonormalization
        const qCols = [];
        for (let c = 0; c < d; c++) {
            const v = [...rawCols[c]];
            for (let prev = 0; prev < c; prev++) {
                const dot = v.reduce((sum, val, idx) => sum + val * qCols[prev][idx], 0);
                for (let r = 0; r < d; r++) {
                    v[r] -= dot * qCols[prev][r];
                }
            }
            const norm = Math.hypot(...v);
            const qCol = v.map((val) => val / (norm || 1));
            qCols.push(qCol);
        }
        for (let i = 0; i < numSymbols; i++) {
            codebook[symbols[i]] = qCols[i];
        }
    }
    else {
        // Over capacity: random unit vectors on S^(d-1)
        for (let i = 0; i < numSymbols; i++) {
            const raw = [];
            for (let r = 0; r < d; r++) {
                raw.push(nextGaussian(rng));
            }
            const norm = Math.hypot(...raw);
            codebook[symbols[i]] = raw.map((val) => val / (norm || 1));
        }
    }
    return codebook;
}

File: core/linear_memory.py
"""
TRACE: Linear Recurrent Memory Core Module
==========================================

Primary Source:
BDH-CQ (Engdahl, Kosowski, Chorowski, Stamirowska et al., 2026, arXiv:2608.09888), §3.2,
which defines the general update S_t = U_θ(S_{t-1}, D_t) and notes its additive special case
S_t = S_{t-1} + U_θ(D_t).

EXPLICIT NON-CLAIM:
This toy model instantiates only the additive special case that BDH-CQ's own paper names
as the linear-attention view of its general update rule (with U_θ(D_t) = v_t k_t^T).
It does not reconstruct U_θ, whose form is proprietary and undisclosed.
It must never be described as reproducing BDH or BDH-CQ.
"""

from typing import Dict, List, Optional, Tuple, Union
import numpy as np


class LinearMemory:
    """
    Fixed-capacity linear associative memory:
        State:    S ∈ R^(d × d), S_0 = 0
        Update:   S_t = S_{t-1} + v_t k_t^T
        Readout:  ŷ = (S_t · k_query) / (k_query · k_query)
        Leak:     Interference(k_a, k_b) = (k_a · k_b) / (k_a · k_a)
    """

    def __init__(self, d: int, carrier: Optional[np.ndarray] = None):
        if d < 1:
            raise ValueError(f"Dimension d must be >= 1, got {d}")
        self.d = d
        self.state = np.zeros((d, d), dtype=np.float64)
        self.write_history: List[Dict[str, Union[str, float, List[float]]]] = []

        # Canonical unit carrier for scalar value encoding/decoding
        if carrier is None:
            c = np.ones(d, dtype=np.float64)
            self.carrier = c / np.linalg.norm(c)
        else:
            c = np.asarray(carrier, dtype=np.float64).flatten()
            if len(c) != d:
                raise ValueError(f"Carrier dimension {len(c)} must match d={d}")
            self.carrier = c / np.linalg.norm(c)

    def reset(self) -> None:
        """Resets the recurrent state to zero and clears write history."""
        self.state = np.zeros((self.d, self.d), dtype=np.float64)
        self.write_history = []

    def encode_scalar_value(self, val: float) -> np.ndarray:
        """Maps a scalar value into R^d along the carrier direction."""
        return float(val) * self.carrier

    def decode_scalar_value(self, vec: np.ndarray) -> float:
        """Projects a vector in R^d onto the carrier direction to recover scalar."""
        return float(np.dot(vec, self.carrier))

    def write(
        self,
        key: np.ndarray,
        value: Union[float, np.ndarray],
        label: Optional[str] = None,
    ) -> np.ndarray:
        """
        Updates recurrent state: S_t = S_{t-1} + v_t k_t^T
        Returns the updated state matrix.
        """
        k = np.asarray(key, dtype=np.float64).flatten()
        if len(k) != self.d:
            raise ValueError(f"Key dimension {len(k)} does not match state dimension {self.d}")

        if np.isscalar(value) or (isinstance(value, np.ndarray) and value.ndim == 0):
            v = self.encode_scalar_value(float(value))
            scalar_val = float(value)
        else:
            v = np.asarray(value, dtype=np.float64).flatten()
            if len(v) != self.d:
                raise ValueError(f"Value dimension {len(v)} does not match state dimension {self.d}")
            scalar_val = self.decode_scalar_value(v)

        # Outer product update: v k^T ∈ R^(d × d)
        delta_s = np.outer(v, k)
        self.state += delta_s

        self.write_history.append({
            "label": label or f"step_{len(self.write_history)}",
            "scalar_value": scalar_val,
            "key": k.tolist(),
            "state_snapshot": self.state.copy().tolist(),
        })

        return self.state.copy()

    def read(self, k_query: np.ndarray) -> np.ndarray:
        """
        Readout vector: ŷ = (S_t · k_query) / (k_query · k_query)
        """
        kq = np.asarray(k_query, dtype=np.float64).flatten()
        if len(kq) != self.d:
            raise ValueError(f"Query key dimension {len(kq)} does not match state dimension {self.d}")

        denom = float(np.dot(kq, kq))
        if denom == 0.0:
            raise ZeroDivisionError("Query key norm is zero.")

        numer = self.state @ kq
        return numer / denom

    def read_scalar(self, k_query: np.ndarray) -> float:
        """Reads out the vector and projects onto the scalar carrier."""
        vec = self.read(k_query)
        return self.decode_scalar_value(vec)

    @staticmethod
    def interference_score(key_target: np.ndarray, key_interferer: np.ndarray) -> float:
        """
        Calculates theoretical leakage coefficient:
            Interference(k_target, k_interferer) = (k_target · k_interferer) / (k_target · k_target)
        """
        ka = np.asarray(key_target, dtype=np.float64).flatten()
        kb = np.asarray(key_interferer, dtype=np.float64).flatten()
        denom = float(np.dot(ka, ka))
        if denom == 0.0:
            raise ZeroDivisionError("Target key norm is zero.")
        return float(np.dot(ka, kb) / denom)


def build_codebook(
    symbols: List[str],
    d: int,
    seed: int = 42,
) -> Dict[str, List[float]]:
    """
    Constructs a codebook mapping symbols to unit key vectors in R^d:
    - If len(symbols) <= d: constructs an orthonormal codebook via QR decomposition.
      Zero cross-key interference by construction.
    - If len(symbols) > d: constructs random unit vectors on S^(d-1).
      Vectors are forced to overlap because exact orthogonality is impossible in R^d.
    """
    num_symbols = len(symbols)
    rng = np.random.default_rng(seed)

    if num_symbols <= d:
        # Construct orthonormal keys via QR decomposition
        gaussian_matrix = rng.standard_normal((d, d))
        q, _ = np.linalg.qr(gaussian_matrix)
        # Each column of Q is an orthonormal unit vector in R^d
        codebook = {symbols[i]: q[:, i].tolist() for i in range(num_symbols)}
    else:
        # Over-capacity: random vectors on S^(d-1)
        raw_vectors = rng.standard_normal((num_symbols, d))
        norms = np.linalg.norm(raw_vectors, axis=1, keepdims=True)
        unit_vectors = raw_vectors / norms
        codebook = {symbols[i]: unit_vectors[i].tolist() for i in range(num_symbols)}

    return codebook
